Caching configuration update (#389)
owncloud/docs-server: modules/admin_manual/pages/configuration/server/caching_configuration.adoc

=== APCu

> The easiest cache to use is APCu, because it is a pure data cache, very fast as it is memory based only and _nothing needs to be configured_. APCu can not be used, when needed to run on an external server like in a distributed environment.

Reviewer: ```suggestion
The easiest cache to use is APCu, because it is a pure data cache, very fast as it is memory-based only and _nothing needs to be configured_. APCu can not be used when running on an external server like in a distributed environment.
```

{redis_url}[Redis] is an excellent modern memory cache to use for both distributed caching and as a local cache for
xref:configuration/files/files_locking_transactional.adoc[transactional file locking], because it guarantees that cached objects are available for as long as they are needed.

The performance of Redis when used with a socket connection can get close to the performance of APCu.

NOTE: The Redis PHP module must be at least version 2.2.6 or higher.

[NOTE]
====
* For Ubuntu 20.04, the default shipped Redis Server version and the php-redis extension is 5.x.
* For Ubuntu 22.04, the default shipped Redis Server version is 6.0.16.
* For Ubuntu 22.04, the php7.4-redis extension has version 5.3.x.

> * With Redis version 6, a new authentication mechanism has been introduced named ACL (Access Control Lists). ownCloud does currently not support Redis ACL´s, but does support the password protection, available with current Redis versions.

Reviewer: ```suggestion
* With Redis version 6, a new authentication mechanism has been introduced named ACL (Access Control Lists). ownCloud does currently not support Redis ACLs, but does support the password protection available with current Redis versions.
```

After that, assuming that you don’t encounter any errors, restart Apache and the extension is ready to use.

==== Redis Performance Consideration

* Because Redis writes data to disk, it is recommended to use a high performance backend like SSD.

> * If you do not have a distributed environment where the Redis server is installed at a different server, you can configure Redis to be xref:redis-configuration-using-unix-sockets[socket based] instead TCP based to xref:redis-performance-comparison-tcp-vs-socket[increase performance].

Reviewer: ```suggestion
* If you do not have a distributed environment where the Redis server is installed at a different machine, you can configure Redis to be xref:redis-configuration-using-unix-sockets[socket-based] instead of TCP-based to xref:redis-performance-comparison-tcp-vs-socket[increase performance].
```

==== Additional notes for Redis vs. APCu on Memory Caching

> * APCu is faster at local caching than Redis when using TCP, as APCu is memory based only.

Reviewer: ```suggestion
* APCu is faster at local caching than Redis when using TCP since APCu is memory-based only.
```

=== Redis Configuration

Redis is very configurable; consult the {redis_doc_url}[Redis documentation] to learn more.

> Regardless of whether you have setup Redis to use TCP or a Unix socket, we recommend adding the following for best performance. This enables xref:configure-transactional-file-locking[External Transactional File Locking] based on Redis:

Reviewer: ```suggestion
Regardless of whether you have set up Redis to use TCP or a Unix socket, we recommend adding the following for best performance. This enables xref:configure-transactional-file-locking[External Transactional File Locking] based on Redis:
```

----
'memcache.local' => '\OC\Memcache\Redis',
'redis' => [
    'host' => 'localhost',  // For a Unix domain socket, use '/var/run/redis/redis.sock'
    'port' => {std-port-redis},         // Set to 0 when using a Unix socket
    'timeout' => 0,         // Optional, keep connection open forever
    'password' => '',       // Optional, if not defined no password will be used.
    'dbindex' => 0,         // Optional, if undefined SELECT will not run and will
                            // use Redis Server's default DB Index.
],
----

==== Redis Configuration Using Unix Sockets

If Redis is running on the same server as ownCloud, it is recommended to configure it to use Unix sockets for increased performance. Follow these steps to do so:

. Make the `redis` user a member of the `www-data` group:
+
[source,bash]
----
sudo usermod -g www-data redis
----

> . Create your redis folder that the unix socket will be in:

Reviewer: ```suggestion
. Create your Redis folder that the Unix socket will be in:
```

+
[source,bash]
----
sudo mkdir -p /var/run/redis/
----

. Set the permissions so the redis user and www-data group own it:
+
[source,bash]
----
sudo chown -R redis:www-data /var/run/redis
----

> . Open an editor of your choice and change/add the following parameters to the redis config. Note that in the example, `vi` is used as editor:

Reviewer: ```suggestion
. Open an editor of your choice and change/add the following parameters to the Redis configuration file. Note that in the example `vi` is used as editor:
```

+
[source,bash]
----
sudo vi /etc/redis/redis.conf
----
+
[source,plaintext]
----
# create a unix domain socket to listen on
unixsocket /var/run/redis/redis.sock

# set permissions for the socket
unixsocketperm 770
----

. Restart Redis to make the changes effective:
+
[source,plaintext]
----
sudo service redis-server restart
----

> . Check if the Redis unix socket was created:

Reviewer: ```suggestion
. Check if the Redis Unix socket was created:
```

+
[source,plaintext]
----
ls -lh /var/run/redis
----
+
The output should be like the following:
+
[source,plaintext]
----
srwxrwx--- 1 redis redis 0 May 12 12:26 redis.sock
----

> Then, configure ownCloud to communicate with Redis socket as in the following example. 

Reviewer: ```suggestion
Then, configure ownCloud to communicate with the Redis socket as in the following example. 
```
